helix-term/src/commands.rs

```diff
@@ -3942,43 +3961,188 @@ pub fn completion(cx: &mut Context) {
-            };
-
-            if !prefix.is_empty() {
-                items.retain(|item| {
-                    item.filter_text
+            }
+            .into_iter()
+            .filter_map(|item| {
+                if !prefix.is_empty()
+                    && !item
+                        .filter_text
                         .as_ref()
                         .unwrap_or(&item.label)
                         .starts_with(&prefix)
-                });
-            }
+                {
+                    None
+                } else {
+                    Some(CompletionItem::LSP {
+                        language_server_id,
+                        item: Box::new(item),
+                        offset_encoding,
+                    })
+                }
+            })
+            .collect::<Vec<_>>();
 
             if items.is_empty() {
                 // editor.set_error("No completion available");
                 return;
             }
             let size = compositor.size();
             let ui = compositor.find::<ui::EditorView>().unwrap();
-            ui.set_completion(
-                editor,
-                items,
-                offset_encoding,
-                start_offset,
-                trigger_offset,
-                size,
-            );
+            ui.set_or_extend_completion(editor, items, start_offset, trigger_offset, size);
         },
     );
 }
 
+pub fn completion_path(cx: &mut Context) {
+    let (view, doc) = current!(cx.editor);
+
+    let text = doc.text().slice(..);
+    let cursor = doc.selection(view.id).primary().cursor(text);
+    let cur_line = text.char_to_line(cursor);
+    let begin_line = text.line_to_char(cur_line);
+    let line_until_cursor = text.slice(begin_line..cursor).to_string();
+
+    // TODO: trigger_offset should be the cursor offset but we also need a starting offset from where we want to apply
+    // completion filtering. For example logger.te| should filter the initial suggestion list with "te".
+
+    use helix_core::chars;
+    let mut iter = text.chars_at(cursor);
+    iter.reverse();
+    let offset = iter.take_while(|ch| chars::char_is_word(*ch)).count();
+
+    let trigger_offset = cursor;
+    let start_offset = cursor.saturating_sub(offset);
+
+    let callback = async move {
+        let call = job::Callback::EditorCompositor(Box::new(
+            move |editor: &mut Editor, compositor: &mut Compositor| {
+                if editor.mode != Mode::Insert {
+                    // we're not in insert mode anymore
+                    return;
+                }
+                let doc = doc!(editor);
+
+                // TODO async path completion (for this probably the whole completion system has to be reworked to be async without producing race conditions)
+                let items = ui::PATH_REGEX
+                    .find(&line_until_cursor)
+                    .and_then(|matched_path| {
+                        let matched_path = matched_path.as_str();
```

Comment:
Would it be possible to also add a denylist setting? I run NixOS, so attempting to index `/nix/store` would be extremely slow with its ~265000 files / directories.

Reply:
Maybe we can just use the  `.(git)ignore` for that?

Actually I've used `nix/store`  for stress testing/general feeling, and I thought it was acceptable given the amount of files (but I've got a fast CPU and NVME drive for /nix ...))

Would someone actually edit something directly in `/nix/store` or in a directory with that much files/folders?

I'm not sure of a daily usecase currently where this would really be a problem, but I'm open for different solutions (my suggestion would be a global .ignore)

Reply:
I don't _edit_ things in the Nix store, no, but I do use Helix to read files in there very frequently. I'd rather not use a .ignore for that because so many other tools read that, and I'd like to set this only in Helix.

Reply:
Right now it's less of a blocker (literally :)) for typing, as it's now async and results will be discarded when not relevant anymore (i.e. the user has typed something that doesn't match the completion item).

On a fast computer it's still *acceptable* IMHO, especially if the folder was indexed once (i.e. is likely cached in memory when accessing again)).

If this is still relevant, can you/someone maybe think/prototype how the UX/configuration for this look like (I'm happy to implement it, but I'm not sure how the UI for this should look like).

But the (completion-)menu is quite slow with > ~100k entries that are sorted synchronously every keystroke.
I think a more sophisticated lazier/async/"streamable" menu/picker implementation would be probably help in general (I still think that the behavior is so similar (also in the future), that the implementations should be shared, but that's a different topic...).